shared-libs/rules-engine/src/rules-state-store.js

```diff
@@ -177,12 +183,41 @@ const self = {
    * @returns {string} User contact document
    */
   currentUser: () => currentUser,
+
+  /**
+   * Store a set of target emissions which were emitted by refreshing a set of contacts
+   *
+   * @param {string[]} contactIds An array of contact ids which produced these targetEmissions by being refreshed. If undefined, all contacts are updated.
+   * @param {Object[]} targetEmissions An array of target emissions (the result of the rules-emitter).
+   */
+  storeTargetEmissions: (contactIds, targetEmissions) => {
+    const isUpdated = targetState.storeTargetEmissions(state.targetState, contactIds, targetEmissions);
+    if (isUpdated) {
+      return onStateChange(state);
+    }
+  },
+
+  /**
+   * Aggregates the stored target emissions into target models
+   *
+   * @param {Function(emission)=} targetEmissionFilter Filter function to filter which target emissions should be aggregated
+   * @example aggregateStoredTargetEmissions(emission => emission.date > moment().startOf('month').valueOf())
+   *
+   * @returns {Object[]} result
+   * @returns {string} result[n].* All attributes of the target as defined in the settings doc
+   * @returns {Integer} result[n].total The total number of unique target emission ids matching instanceFilter
+   * @returns {Integer} result[n].pass The number of unique target emission ids matching instanceFilter with the latest emission with truthy "pass"
+   * @returns {Integer} result[n].percent The percentage of pass/total
+   */
+  aggregateStoredTargetEmissions: targetEmissionFilter => targetState.aggregateStoredTargetEmissions(state.targetState, targetEmissionFilter),
 };
 
 const hashRulesConfig = (settingsDoc, userDoc) => {
   const rulesConfig = {
     rules: settingsDoc && settingsDoc.tasks && settingsDoc.tasks.rules,
-    targets: settingsDoc && settingsDoc.targets,
+    targets: settingsDoc && settingsDoc.tasks && settingsDoc.tasks.targets,
+    can_view_tasks: settingsDoc && settingsDoc.permissions && settingsDoc.permissions.can_view_tasks,
+    can_view_analytics: settingsDoc && settingsDoc.permissions && settingsDoc.permissions.can_view_analytics,
```

Comment:
There's a whole lot of repetition in here. What about...

```js
const settingsTasks = settingsDoc && settingsDoc.tasks || {};
const settingsPermissions = settingsDoc && settingsDoc.permissions || {};
const rulesConfig = {
  rules: settingsTasks.rules,
  targets: settingsTasks.targets,
  ...
  userDoc,
}
```

You could also do something clever with destructuring and Object.assign but I'll leave that as an exercise for the reader...

Reply:
No need for destructuring I don't think....